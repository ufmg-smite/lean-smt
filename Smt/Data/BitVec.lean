/-
Copyright (c) 2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Joe Hendrix, Wojciech Nawrocki
-/

/-!
We define bitvectors in two variants - indexed and packed. The indexed variant is helpful
for stating strongly-typed interfaces, whereas the packed one is better for stating some
properties without the dependent index getting in the way.

TODO(WN): explain why the particular choice of definition

TODO(WN): This is planned to go into mathlib4 once we:
  - prove the various bounds
  - match the interface to what SMT-LIB provides
  - are otherwise happy with the API
-/

def BitVec (w : Nat) := Fin (2^w)

structure BitVec.Packed where
  width : Nat
  data  : Fin (2^width)

namespace BitVec
namespace Packed

theorem ext {a b : Packed} (hWidth : a.width = b.width)
    (hData : a.data.val = b.data.val) : a = b := by
  let ⟨aw, ad, _⟩ := a
  let ⟨bw, bd, _⟩ := b
  cases hWidth
  cases hData
  rfl

end Packed

protected def zero (w : Nat) : BitVec w :=
  ⟨0, Nat.pos_pow_of_pos _ <| by decide⟩

instance : Inhabited (BitVec w) := ⟨BitVec.zero w⟩

instance : OfNat (BitVec w) (nat_lit 0) :=
  ⟨BitVec.zero w⟩

protected def append (x : BitVec w) (y : BitVec v) : BitVec (w+v) :=
  ⟨x.val <<< v ||| y.val, sorry⟩

instance : HAppend (BitVec w) (BitVec v) (BitVec (w+v)) where
  hAppend := BitVec.append

protected def and (x y : BitVec w) : BitVec w :=
  ⟨x.val &&& y.val, sorry⟩

protected def or (x y : BitVec w) : BitVec w :=
  ⟨x.val ||| y.val, sorry⟩

protected def xor (x y : BitVec w) : BitVec w :=
  ⟨x.val ^^^ y.val, sorry⟩

protected def shiftLeft (x : BitVec w) (n : Nat) : BitVec w :=
  Fin.ofNat' (x.val <<< n) (Nat.pos_pow_of_pos _ (by decide))

protected def shiftRight (x : BitVec w) (n : Nat) : BitVec w :=
  ⟨x.val >>> n, sorry⟩

instance : AndOp (BitVec w) := ⟨BitVec.and⟩
instance : OrOp (BitVec w) := ⟨BitVec.or⟩
instance : Xor (BitVec w) := ⟨BitVec.xor⟩
instance : HShiftLeft (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftLeft⟩
instance : HShiftRight (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftRight⟩

-- TODO re-express as extract == #b0
def lsb_get! (x : BitVec m) (i : Nat) : Bool :=
  (x.val &&& (1 <<< i)) ≠ 0

-- TODO re-express
def lsb_set! (x : BitVec m) (i : Nat) (c : Bool) : BitVec m :=
  if c then
    x ||| ⟨1 <<< i, sorry⟩
  else
    x &&& ⟨((1 <<< m) - 1 - (1 <<< i)), sorry⟩

-- TODO how to zero-extend in SMT-LIB?
def zeroShrinxtend (v : Nat) (x : BitVec w) : BitVec v :=
  if h : w < v then ⟨x.val, Nat.lt_trans x.isLt sorry⟩
  else ⟨x.val % (2^v), Nat.mod_lt _ sorry⟩

end BitVec
