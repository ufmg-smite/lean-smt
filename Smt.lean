import Smt.General
import Smt.Nat
import Smt.Prop
import Smt.Solver
import Smt.String
import Smt.Tactic
import Smt.Term
