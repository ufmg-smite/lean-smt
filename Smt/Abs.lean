/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Smt.Translator
import Mathlib.Algebra.Order.Group.Abs

namespace Smt.Abs

open Lean Expr
open Translator Term


@[smtTranslator] def replaceAbs : Translator
  | app (app (app (app (const ``abs _) _) _) _) e => do
      let tmE ← applyTranslators! e
      return appT (symbolT "abs") tmE
  | _ => return none

end Smt.Abs
