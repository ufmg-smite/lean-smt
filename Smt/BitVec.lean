/-
Copyright (c) 2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Wojciech Nawrocki
-/

import Lean
import Smt.Translator
import Smt.Data.BitVec

namespace Smt.BitVec

open Lean Expr
open Translator Term

@[smtTranslator] def replaceType : Translator
  | e@(app (const ``BitVec _) n) => do
    let n ← Meta.whnf n
    let some n ← Meta.evalNat n |>.run
      | throwError "bitvector type{indentD e}\nhas variable width"
    return mkApp2 (symbolT "_") (symbolT "BitVec") (literalT (toString n))
  | _ => return none

def mkLit (w : Nat) (n : Nat) : Term :=
  let bits := Nat.toDigits 2 n |>.take w
  literalT <| bits.foldl (init := "#b".pushn '0' (w - bits.length)) (·.push ·)

@[smtTranslator] def replaceEq : Translator
  -- TODO: we should really just support beq/bne across all types uniformly
  | app (app (const ``BEq.beq _) (app (const ``BitVec _) _)) _ => return symbolT "="
  | app (app (app (app (const ``bne _) (app (const ``BitVec _) _)) _) e) e' =>
    return appT (symbolT "not") (mkApp2 (symbolT "=") (← applyTranslators! e) (← applyTranslators! e'))
  | _ => return none

@[smtTranslator] def replaceFun : Translator
  | app (app (const ``BitVec.append _) _) _ => return symbolT "concat"
  | app (const ``BitVec.and _) _            => return symbolT "bvand"
  | app (const ``BitVec.or _) _             => return symbolT "bvor"
  | app (const ``BitVec.xor _) _            => return symbolT "bvxor"
  | e@(app (app (app (const ``BitVec.shiftLeft _) w) x) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkApp2 (symbolT "bvshl") (← applyTranslators! x) (mkLit w n)
  | e@(app (app (app (const ``BitVec.shiftRight _) w) x) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkApp2 (symbolT "bvlshr") (← applyTranslators! x) (mkLit w n)
  | e@(app (app (app (const ``BitVec.extract _) _) i) j) => do
    let i ← reduceLit i e
    let j ← reduceLit j e
    return mkApp3 (symbolT "_") (symbolT "extract") (literalT (toString i)) (literalT (toString j))
  | e@(app (const ``BitVec.zero _) w) => do
    let w ← reduceWidth w e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkLit w 0
  | e@(app (app (const ``BitVec.ofNat _) w) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkLit w n
  | _ => return none
where
  reduceWidth (w : Expr) (e : Expr) : TranslationM Nat := do
    let some w ← Meta.evalNat (← Meta.whnf w) |>.run
      | throwError "bitvector width{indentD w}\nis not constant in{indentD e}"
    return w
  reduceLit (n : Expr) (e : Expr) : TranslationM Nat := do
    let some n ← Meta.evalNat (← Meta.whnf n) |>.run
      | throwError "literal{indentD n}\nis not constant in{indentD e}"
    return n

end Smt.BitVec
