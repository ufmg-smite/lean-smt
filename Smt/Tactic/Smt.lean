/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean

import Smt.Dsl.Sexp
import Smt.Query
import Smt.Solver

namespace Smt

open Lean
open Lean.Elab
open Lean.Elab.Tactic

initialize
  registerTraceClass `smt.debug
  registerTraceClass `smt.debug.attr
  registerTraceClass `smt.debug.query
  registerTraceClass `smt.debug.transformer

/-- `smt` converts the current goal into an SMT query and checks if it is
satisfiable. By default, `smt` generates the minimum valid SMT query needed to
assert the current goal. However, that is not always enough:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt
```
For the theorem above, `smt` generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(check-sat)
```
which is missing the hypotheses `hp` and `hpq` required to prove the theorem. To
pass hypotheses to the solver, use `smt [h₁, h₂, ..., hₙ]` syntax:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt [hp, hpq]
```
The tactic then generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(declare-const p Bool)
(assert p)
(assert (=> p q))
(check-sat)
```
-/
syntax (name := smt) "smt" ("[" ident,* "]")? : tactic

def queryToString (commands : List String) : String :=
  String.intercalate "\n" ("(check-sat)\n" :: commands).reverse

def parseTactic : Syntax → TacticM (List Expr)
  | `(tactic| smt)              => pure []
  | `(tactic| smt [ $[$hs],* ]) => hs.toList.mapM (fun h => elabTerm h.raw none)
  | _                           => throwUnsupportedSyntax

@[tactic smt] def evalSmt : Tactic := fun stx => withMainContext do
  -- 1. Get the current main goal.
  let goalType ← Tactic.getMainTarget
  let goalId ← Lean.mkFreshMVarId
  Lean.Meta.withLocalDeclD goalId.name (mkNot goalType) fun g => do
  -- 2. Get the hints passed to the tactic.
  let mut hs ← parseTactic stx
  hs := hs.eraseDups
  -- 3. Generate the SMT query.
  let mut solver := Solver.mk []
  solver ← Query.generateQuery g hs solver
  let query := queryToString solver.commands
  -- 4. Run the solver.
  let kind := smt.solver.kind.get (← getOptions)
  let path := smt.solver.path.get? (← getOptions) |>.getD (toString kind)
  -- 5. Print the result.
  let res ← solver.checkSat kind path
  logInfo m!"goal: {goalType}\n\nquery:\n{query}\nresult: {res}"
  let out ← match Sexp.parse res with
    | .ok out => pure out
    | .error e => throwError "cannot parse solver output: {e}"
  if out.contains sexp!{sat} then
    throwError "unable to prove goal, it is false"
    -- TODO ask for countermodel and print
  else if !out.contains sexp!{unsat} then
    throwError "unable to prove goal"

end Smt
