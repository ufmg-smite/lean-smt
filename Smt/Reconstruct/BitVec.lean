/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Smt.Reconstruct.Bool.Tactic
import Smt.Reconstruct.BitVec.Bitblast
import Smt.Reconstruct.Prop.Core
import Smt.Reconstruct

def Std.Range.foldlM [Monad m] (f : α → Nat → m α) (r : Range) (init : α) : m α := do
  let mut a := init
  for i in r do
    a ← f a i
  return a

namespace Smt.Reconstruct.BitVec

open Lean Qq

@[smt_sort_reconstruct] def reconstructBitVecSort : SortReconstructor := fun s => do match s.getKind with
  | .BITVECTOR_SORT =>
    let w : Nat := s.getBitVectorSize!.val
    return q(BitVec $w)
  | _             => return none

partial def synthDecidableInst (t : cvc5.Term) : ReconstructM Expr := do match t.getKind with
    | .CONST_BOOLEAN => return if t.getBooleanValue! then q(instDecidableTrue) else q(instDecidableFalse)
    | .NOT =>
      let p : Q(Prop) ← reconstructTerm t[0]!
      let hp : Q(Decidable $p) ← synthDecidableInst t[0]!
      return q(@instDecidableNot $p $hp)
    | .AND => rightAssocOpDecidableInst q(And) q(@instDecidableAnd) t
    | .OR => rightAssocOpDecidableInst q(Or) q(@instDecidableOr) t
    | .XOR => rightAssocOpDecidableInst q(XOr) q(@XOr.instDecidable) t
    | .EQUAL =>
      if t[0]!.getSort.getKind == .BOOLEAN_SORT then
        let p : Q(Prop) ← reconstructTerm t[0]!
        let q : Q(Prop) ← reconstructTerm t[1]!
        let hp : Q(Decidable $p) ← synthDecidableInst t[0]!
        let hq : Q(Decidable $q) ← synthDecidableInst t[1]!
        return q(@instDecidableEqOfIff $p $q (@instDecidableIff $p $q $hp $hq))
      if t[0]!.getSort.getKind == .BITVECTOR_SORT then
        let w : Nat := t[0]!.getSort.getBitVectorSize!.val
        return q(@instDecidableEqBitVec $w)
      let p : Q(Prop) ← reconstructTerm t
      Meta.synthInstance q(Decidable $p)
    | .BITVECTOR_BIT =>
      let w : Nat := t[0]!.getSort.getBitVectorSize!.val
      let x : Q(BitVec $w) ← reconstructTerm t[0]!
      let i : Nat := t.getOp![0]!.getIntegerValue!.toNat
      return q(instDecidableEqBool («$x».getLsbD $i) true)
    | _ =>
      let p : Q(Prop) ← reconstructTerm t
      Meta.synthInstance q(Decidable $p)
where
   rightAssocOpDecidableInst (op : Expr) (inst : Expr) (t : cvc5.Term) : ReconstructM Expr := do
    let mut curr ← reconstructTerm t[t.getNumChildren - 1]!
    let mut currInst ← synthDecidableInst t[t.getNumChildren - 1]!
    for i in [1:t.getNumChildren] do
      let ct := t[t.getNumChildren - i - 1]!
      currInst := mkApp4 inst (← reconstructTerm ct) curr (← synthDecidableInst ct) currInst
      curr := mkApp2 op (← reconstructTerm ct) curr
    return currInst

@[smt_term_reconstruct] def reconstructBitVec : TermReconstructor := fun t => do match t.getKind with
  | .CONST_BITVECTOR =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let v : Nat := (t.getBitVectorValue! 10).toNat!
    return q(BitVec.ofNat $w $v)
  | .BITVECTOR_CONCAT =>
    let n : Nat := t.getNumChildren
    let w₁ : Nat := t[0]!.getSort.getBitVectorSize!.val
    let a : Q(BitVec $w₁) ← reconstructTerm t[0]!
    let f := fun ⟨w₁, a⟩ i => do
      let w₂ : Nat := t[i]!.getSort.getBitVectorSize!.val
      let x : Q(BitVec $w₂) ← reconstructTerm t[i]!
      return ⟨q($w₁ + $w₂), q($a ++ $x)⟩
    let ⟨_, a⟩ ← [1:n].foldlM f (⟨q($w₁), a⟩ : Σ w : Q(Nat), Q(BitVec $w))
    return a
  | .BITVECTOR_AND =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    leftAssocOp q(@AndOp.and (BitVec $w) _) t
  | .BITVECTOR_OR =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    leftAssocOp q(@OrOp.or (BitVec $w) _) t
  | .BITVECTOR_XOR =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    leftAssocOp q(@Xor.xor (BitVec $w) _) t
  | .BITVECTOR_NOT =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    return q(~~~$x)
  | .BITVECTOR_MULT =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    leftAssocOp q(@HMul.hMul (BitVec $w) (BitVec $w) (BitVec $w) _) t
  | .BITVECTOR_ADD =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    leftAssocOp q(@HAdd.hAdd (BitVec $w) (BitVec $w) (BitVec $w) _) t
  | .BITVECTOR_SUB =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    leftAssocOp q(@HSub.hSub (BitVec $w) (BitVec $w) (BitVec $w) _) t
  | .BITVECTOR_NEG =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    return q(-$x)
  | .BITVECTOR_UDIV =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q(BitVec.smtUDiv $x $y)
  | .BITVECTOR_UREM =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q($x % $y)
  | .BITVECTOR_SDIV =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q(BitVec.smtSDiv $x $y)
  | .BITVECTOR_SREM =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q(BitVec.srem $x $y)
  | .BITVECTOR_SMOD =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q(BitVec.smod $x $y)
  | .BITVECTOR_SHL =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let i : Nat := t.getOp![0]!.getIntegerValue!.toNat
    return q($x <<< $i)
  | .BITVECTOR_LSHR =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let i : Nat := t.getOp![0]!.getIntegerValue!.toNat
    return q($x >>> $i)
  | .BITVECTOR_ASHR =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let i : Nat := t.getOp![0]!.getIntegerValue!.toNat
    return q(BitVec.sshiftRight $x $i)
  | .BITVECTOR_ULT =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q($x < $y)
  | .BITVECTOR_ULE =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q($x ≤ $y)
  | .BITVECTOR_UGT =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q($x > $y)
  | .BITVECTOR_UGE =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q($x ≥ $y)
  | .BITVECTOR_SLT =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q(BitVec.slt $x $y)
  | .BITVECTOR_SLE =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let y : Q(BitVec $w) ← reconstructTerm t[1]!
    return q(BitVec.sle $x $y)
  | .BITVECTOR_EXTRACT =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let i : Nat := t.getOp![0]!.getIntegerValue!.toNat
    let j : Nat := t.getOp![1]!.getIntegerValue!.toNat
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    return q(«$x».extractLsb $i $j)
  | .BITVECTOR_REPEAT =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let n : Nat := t.getOp![0]!.getIntegerValue!.toNat
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    return q(«$x».replicate $n)
  | .BITVECTOR_ZERO_EXTEND =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let n : Nat := t.getOp![0]!.getIntegerValue!.toNat
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    return q(«$x».zeroExtend $n)
  | .BITVECTOR_SIGN_EXTEND =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let n : Nat := t.getOp![0]!.getIntegerValue!.toNat
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    return q(«$x».signExtend $n)
  | .BITVECTOR_ROTATE_LEFT =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let n : Nat := t.getOp![0]!.getIntegerValue!.toNat
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    return q(«$x».rotateLeft $n)
  | .BITVECTOR_ROTATE_RIGHT =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let n : Nat := t.getOp![0]!.getIntegerValue!.toNat
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    return q(«$x».rotateRight $n)
  | .INT_TO_BITVECTOR =>
    let w : Nat := t.getSort.getBitVectorSize!.val
    let x : Q(Int) ← reconstructTerm t[0]!
    return q(BitVec.ofNat $w «$x».toNat)
  | .BITVECTOR_TO_NAT =>
    let w : Nat := t[0]!.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    return q(«$x».toNat)
  | .BITVECTOR_FROM_BOOLS =>
    let w : Nat := t.getNumChildren
    let bs : Q(BitVec 0) := q(.nil)
    let f (bs : Expr) (i : Nat) : ReconstructM Expr := do
      let p : Q(Prop) ← reconstructTerm t[i]!
      let bs : Q(BitVec $i) := bs
      let hp : Q(Decidable $p) ← synthDecidableInst t[i]!
      return q(@BitVec.cons $i (@decide $p $hp) $bs)
    let bs : Q(BitVec $w) ← (List.range w).foldlM f bs
    return q($bs)
  | .BITVECTOR_BIT =>
    let w : Nat := t[0]!.getSort.getBitVectorSize!.val
    let x : Q(BitVec $w) ← reconstructTerm t[0]!
    let i : Nat := t.getOp![0]!.getIntegerValue!.toNat
    return q(«$x».getLsbD $i = true)
  | _ => return none
where
  leftAssocOp (op : Expr) (t : cvc5.Term) : ReconstructM Expr := do
    let mut curr ← reconstructTerm t[0]!
    for i in [1:t.getNumChildren] do
      curr := mkApp2 op curr (← reconstructTerm t[i]!)
    return curr

def reconstructRewrite (pf : cvc5.Proof) : ReconstructM (Option Expr) := do
  match pf.getRewriteRule with
  | _ => return none

@[smt_proof_reconstruct] def reconstructBitVecProof : ProofReconstructor := fun pf => do match pf.getRule with
  | .DSL_REWRITE => reconstructRewrite pf
  | .BV_BITBLAST_STEP =>
    let t := pf.getArguments[0]![0]!
    match t.getKind with
    | .CONST_BITVECTOR =>
      let w : Nat := t.getSort.getBitVectorSize!.toNat
      let t : Q(BitVec $w) ← reconstructTerm pf.getResult[0]!
      let t' : Q(BitVec $w) ← reconstructTerm pf.getResult[1]!
      addThm q($t = $t') q(Eq.refl $t)
    | .CONSTANT =>
      let w : Nat := t.getSort.getBitVectorSize!.toNat
      let x : Q(BitVec $w) ← reconstructTerm pf.getResult[0]!
      let x' : Q(BitVec $w) ← reconstructTerm pf.getResult[1]!
      let h : Q(«$x».bb = $x') ← Meta.mkFreshExprMVar q(«$x».bb = $x')
      let mv ← Bool.boolify h.mvarId!
      let ds := [``BitVec.bb, ``BitVec.iunfoldr, ``Fin.hIterate, ``Fin.hIterateFrom]
      let ps := [``Nat.reduceAdd, ``Nat.reduceLT, ``reduceDIte]
      let simpTheorems ← ds.foldrM (fun n a => a.addDeclToUnfold n) {}
      let simpProcs ← ps.foldrM (fun n a => a.add n false) {}
      let ctx ← Meta.Simp.mkContext {} #[simpTheorems]
      let (some mv, _) ← Meta.simpTarget mv ctx simpProcs | throwError "simp failed"
      mv.refl
      addThm q($x = $x') q(Eq.trans (BitVec.self_eq_bb $x) $h)
    | .EQUAL =>
      let w : Nat := t[0]!.getSort.getBitVectorSize!.toNat
      let x : Q(BitVec $w) ← reconstructTerm pf.getResult[0]![0]!
      let y : Q(BitVec $w) ← reconstructTerm pf.getResult[0]![1]!
      let p : Q(Prop) ← reconstructTerm pf.getResult[1]!
      let hp : Q(Decidable $p) ← synthDecidableInst pf.getResult[1]!
      let h : Q(decide (BitVec.beq $x $y) = decide $p) ← Meta.mkFreshExprMVar q(decide (BitVec.beq $x $y) = @decide $p $hp)
      let mv ← Bool.boolify h.mvarId!
      let ds := [``BitVec.beq, ``BitVec.beq.go]
      let ts := [``BitVec.getLsbD_cons, ``Nat.succ.injEq]
      let ps := [``Nat.reduceAdd, ``Nat.reduceSub, ``Nat.reduceEqDiff, ``reduceIte]
      let simpTheorems ← ds.foldrM (fun n a => a.addDeclToUnfold n) {}
      let simpTheorems ← ts.foldrM (fun n a => a.addConst n) simpTheorems
      let ctx ← Meta.Simp.mkContext {} #[simpTheorems]
      let simpProcs ← ps.foldrM (fun n a => a.add n false) {}
      let (some mv, _) ← Meta.simpTarget mv ctx simpProcs | throwError "simp failed"
      mv.refl
      addThm q(($x = $y) = $p) q(Eq.trans (BitVec.eq_eq_beq $x $y) (Bool.eq_of_decide_eq $h))
    | .BITVECTOR_ADD =>
      let w : Nat := t[0]!.getSort.getBitVectorSize!.toNat
      let x : Q(BitVec $w) ← reconstructTerm pf.getResult[0]![0]!
      let y : Q(BitVec $w) ← reconstructTerm pf.getResult[0]![1]!
      let z : Q(BitVec $w) ← reconstructTerm pf.getResult[1]!
      let h : Q((BitVec.adc' $x $y false).snd = $z) ← Meta.mkFreshExprMVar q((BitVec.adc' $x $y false).snd = $z)
      let mv ← Bool.boolify h.mvarId!
      let ds := [``BitVec.adc', ``BitVec.adcb', ``BitVec.iunfoldr, ``Fin.hIterate, ``Fin.hIterateFrom]
      let ts := [``BitVec.getLsbD_cons, ``Nat.succ.injEq]
      let ps := [``Nat.reduceAdd, ``Nat.reduceLT, ``reduceDIte, ``reduceIte]
      let simpTheorems ← ds.foldrM (fun n a => a.addDeclToUnfold n) {}
      let simpTheorems ← ts.foldrM (fun n a => a.addConst n) simpTheorems
      let ctx ← Meta.Simp.mkContext {} #[simpTheorems]
      let simpProcs ← ps.foldrM (fun n a => a.add n false) {}
      let (some mv, _) ← Meta.simpTarget mv ctx simpProcs | throwError "simp failed"
      mv.refl
      addThm q($x + $y = $z) q(Eq.trans (BitVec.add_eq_adc' $x $y) $h)
    | _ => return none
  | _ => return none

end Smt.Reconstruct.BitVec
