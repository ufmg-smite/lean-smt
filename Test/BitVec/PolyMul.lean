import Smt
import Smt.Data.BitVec

def polyAdd (a b : BitVec w) : BitVec w := a ^^^ b

/-- Long multiplication in GF(2)[x] translated from Cryptol reference.

```cryptol
pmult : {u, v} (fin u, fin v) => [1 + u] -> [1 + v] -> [1 + u + v]
pmult x y = last zs
  where
    zs = [0] # [ (z << 1) ^ (if yi then 0 # x else 0) | yi <- y | z <- zs ]
``` -/
def polyMul (a : BitVec w) (b : BitVec v) : BitVec (w+v) :=
  let wOut := w + v
  let a := a.zeroShrinxtend wOut
  -- fold over the bits of b starting at MSB
  let ret : BitVec wOut := List.range v |>.foldr (init := 0) fun i acc =>
    let acc' := acc <<< 1
    if b.lsb_get! i then polyAdd acc' a else acc'
  ret

theorem polyMul_comm_8 (x y z w : BitVec 2) : polyAdd (polyMul x y) (polyMul z w) = polyAdd (polyMul w z) (polyMul y x)  := by
  concretize [polyAdd, polyMul]
  smt [polyAdd.«4», polyMul.«2».«2»]
  sorry
