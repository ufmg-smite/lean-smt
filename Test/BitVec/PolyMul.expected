Test/BitVec/PolyMul.lean:25:2: error: cannot translate fun i acc =>
  if BitVec.lsb_get! b i = true then polyAdd.«4» (BitVec.shiftLeft acc 1) (BitVec.zeroShrinxtend (Nat.add 2 2) a)
  else BitVec.shiftLeft acc 1, SMT-LIB does not support lambdas
